src/unix/pty.cc

```diff
@@ -131,6 +132,131 @@ pty_after_waitpid(uv_async_t *);
+  sigset_t newmask, oldmask;
+  struct sigaction sig_action;
+
+  // temporarily block all signals
+  // this is needed due to a race condition in openpty
+  // and to avoid running signal handlers in the child
+  // before exec* happened
+  sigfillset(&newmask);
+  pthread_sigmask(SIG_SETMASK, &newmask, &oldmask);
+
+  pid = pty_forkpty(&pty_master, nullptr, term, winp);
+
+  if (!pid) {
+    // remove all signal handler from child
+    sig_action.sa_handler = SIG_DFL;
+    sig_action.sa_flags = 0;
+    sigemptyset(&sig_action.sa_mask);
+    for (int i = 0 ; i < NSIG ; i++) {    // NSIG is a macro for all signals + 1
+      sigaction(i, &sig_action, NULL);
+    }
+  }
+  // reenable signals
+  pthread_sigmask(SIG_SETMASK, &oldmask, NULL);
+
+  if (pid) {
+    if (pid == -1) {
+      Nan::ThrowError("forkpty(3) failed.");
+      return 1;
+    }
+
+    return 0;
+  } else {
+    if (strlen(cwd)) {
+      if (chdir(cwd) == -1) {
+        perror("chdir(2) failed.");
+        _exit(1);
+      }
+    }
+
+    if (uid != -1 && gid != -1) {
+      if (setgid(gid) == -1) {
+        perror("setgid(2) failed.");
+        _exit(1);
+      }
+      if (setuid(uid) == -1) {
+        perror("setuid(2) failed.");
+        _exit(1);
+      }
+    }
+
+    pty_execvpe(argv[0], argv, env);
+
+    perror("execvp(3) failed.");
+    _exit(1);
+  }
+  return 0;
+}
+
+int pty_fork_posix_spawn(char **argv, char **env, const termios *term, winsize *winp, int &pty_master, pid_t &pid) {
+  sigset_t all_signals, oldmask;
+
+  // temporarily block all signals
+  // this is needed due to a race condition in openpty
+  // and to avoid running signal handlers in the child
+  // before exec* happened
+  sigfillset(&all_signals);
+  pthread_sigmask(SIG_SETMASK, &all_signals, &oldmask);
+
+  int pty_slave;
+  if (pty_openpty(&pty_master, &pty_slave, nullptr, term, winp) == -1) {
+    pthread_sigmask(SIG_SETMASK, &oldmask, NULL);
+    Nan::ThrowError("pty_openpty failed.");
+    return 1;
+  }
+
+  pthread_sigmask(SIG_SETMASK, &oldmask, NULL);
+
+  posix_spawn_file_actions_t acts;
+  posix_spawnattr_t attrs;
+  if (posix_spawn_file_actions_init(&acts)) {
+    Nan::ThrowError("posix_spawn_file_actions_init failed.");
+    return 1;
+  }
+  if (posix_spawnattr_init(&attrs)) {
+    Nan::ThrowError("posix_spawnattr_init failed.");
+    return 1;
+  }
+
+  if (
+    posix_spawn_file_actions_adddup2(&acts, pty_slave, STDIN_FILENO) ||
+    posix_spawn_file_actions_adddup2(&acts, pty_slave, STDOUT_FILENO) ||
+    posix_spawn_file_actions_addclose(&acts, pty_slave) ||
+    posix_spawn_file_actions_addclose(&acts, pty_master)
+  ) {
+    posix_spawn_file_actions_destroy(&acts);
+    posix_spawnattr_destroy(&attrs);
+    Nan::ThrowError("posix_spawn_file_actions_init failed.");
+    return 1;
+  }
+
+  if (
+    posix_spawnattr_setsigdefault(&attrs, &all_signals) ||
+    posix_spawnattr_setsigmask(&attrs, &oldmask) ||
```

Comment:
Are you sure this does the right thing? Why would you apply oldmask here for a newly created process? I think those should be all cleared to default handler here, though I am not quite sure about that, there might be typical OS pattern here (prolly a lookup in shell process spawning code might give some clues). The fork way clears them all currently.

Reply:
I was just following the [original fork code](https://github.com/microsoft/node-pty/blob/main/src/unix/pty.cc#L257) where it's unconditionally reapplied after fork - I just assumed we want to inherit the signal settings?

Reply:
Right, the old code does the same. Cant remember why I did this way there. If you feel like doing more C - a smaller test program in C spawned at the pty could test, if the signals are all active with the default handler.

Reply:
Yep - a signal masked in parent is still masked in a forked child. I suppose this was never a problem since there's no way to mask signals in Node anyhow (except from a native module).

Reply:
Hmm, I'd call this a hidden bug tbh, that no one trapped into yet.

Imho a clean way to handle this would be to treat the parent's signal handling as irrelevant for the child, thus spawn it with typical OS defaults. Thats even more true here, since the child shall form a new session behind its own TTY with own process group, which is quite different to normal subprocessing. As written above idk how to derive those OS defaults, and prolly was the reason why I simply inherited it in the fork path (excuses excuses :smile_cat:).

Idea to overcome this - an independent small helper written in C as shell replacement could test and output:
- signal states
- setsid / process group state
- controlling terminal state (whether ioctl TIOCSCTTY needs to be invoked)
- correctness of stdin|stdout|stderr
- leaked other fds
- gid/uid state

This helper directly called in `pty.spawn` then can test the different spawn implementations. Furthermore it can be used to check for common / OS typical values at other terminal emulators. It would also clarify the SETSID and the fd situation, and whether changing the user works.

Reply:
Do we even need all those checks then if the helper can just execve() itself into a new process?

Reply:
Nope, we could just fix all things in that helper and exec into green field. Well, exec errors itself should be propagates backwards somehow, maybe by some messaging pipe, that would CLOEXEC automatically on the final exec step.

Reply:
Done: #487. I've started with using posix_spawn directly to spawn the helper to avoid passing stuff (pipe FD) over `argv`. All parameters go through the pipe as well. 

`posix_spawn` seems fine here, as no non-standard extension APIs are used - just passing two FDs and CLOEXEC'ing the rest.

Let me know if you'd rather have it the other way.